### PR TITLE
Refactor: use wallet RPC when possible; init Safe Core SDK on demand

### LINKED ISSUE #2000: Investigate always using wallet provider when connected
## What is the feature about

We are currently using a mixture of our "read-only" provider and that of the wallet (when it is connected). This can result in two sources of truth if they have different RPCs.

We should investigate how other DApps approach this and whether we should _only_ use that of the wallet if it is connected.

## The list of requirements

- Investigate how other DApps approach this
- Look into where we use which provider for what calls
- Perhaps adjust [provider "creation"](https://github.com/safe-global/safe-wallet-web/blob/dev/src/hooks/wallets/web3.ts#L17-L30) and usage
- Update SDK usage accordingly

## Hypthesised provider

1. Is no wallet connected or UI is on different chain from wallet? -> Use "read-only" provider
2. Is wallet connected and UI on the chain of the wallet? -> Use waller provider

Note: If a wallet provider RPC wasn't working, we would need to fallback to the "read-only" one/notify the user of this otherwise the UI would not work.

## Links

- [Comment](https://github.com/5afe/safe-support/issues/189#issuecomment-1543434882) which raised discussion
- [Discussion](https://5afe.slack.com/archives/C03DAGWJCR5/p1684329486507859)

### PR DESCRIPTION
**Pull request by katspaugh**
## What it solves

Resolves #2000

Another take on #2276

## How this PR fixes it

1. This PR makes INFURA_TOKEN an optional dependency. If it's not defined, and no alternative RPC is configured, a read-only provider will not be created.
2. Same for the Safe Apps provider.
3. `useInitWeb3` will now use the wallet RPC as a read-only provider if it's connected to the right chain
4. ~`useInitSafeCoreSDK` will be called only when necessary, not when the app starts. I.e. in the tx flow. This will reduce the number of initial RPC requests.~

## How to test it
* Create different types of transactions with the wallet connected to the **wrong** chain
* Create different types of transactions with the wallet connected to the **right** chain
* Set a custom RPC, remove a custom RPC

### CODE REVIEW

File: src/hooks/wallets/useInitWeb3.ts [hunk 1]
```diff
@@ -6,27 +6,33 @@ import { createWeb3, createWeb3ReadOnly, setWeb3, setWeb3ReadOnly } from '@/hook
...
-  }, [wallet])
+    if (wallet) {
+      const web3 = createWeb3(wallet.provider)
+      setWeb3(web3)
 
-  useEffect(() => {
-    if (!chain) {
-      return
+      if (wallet.chainId === chain.chainId) {
+        setWeb3ReadOnly(web3)
```
[the remark this answers is not in the document]
Comment (katspaugh) [reply src/hooks/wallets/useInitWeb3.ts:25]: The provider itself might be either read-only (when no wallet), or not read-only, but the places where it's used require only a read-only one.

File: src/hooks/wallets/useInitWeb3.ts [hunk 2]
```diff
@@ -6,27 +6,33 @@ import { createWeb3, createWeb3ReadOnly, setWeb3, setWeb3ReadOnly } from '@/hook
...
 
-    const web3ReadOnly = createWeb3ReadOnly(chain.rpcUri, customRpc?.[chain.chainId])
-    setWeb3ReadOnly(web3ReadOnly)
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [chain?.chainId, customRpc])
+    // Wait for wallet to be connected
+    const timeout = setTimeout(() => {
+      const web3ReadOnly = createWeb3ReadOnly(chain.rpcUri, customRpcUrl)
+      setWeb3ReadOnly(web3ReadOnly)
+    }, READONLY_WAIT)
```
[the remark this answers is not in the document]
Comment (katspaugh) [reply src/hooks/wallets/useInitWeb3.ts:34]: Does it have an event along the lines of "everything is initialized but a wallet is legit not connected, proceed with a read only one"?

[the remark this answers is not in the document]
Comment (iamacook) [reply src/hooks/wallets/useInitWeb3.ts:34]: A wallet is only present in the [`wallets` state](https://onboard.blocknative.com/docs/modules/core#subscribe-to-state-updates) if it is connected.

[the remark this answers is not in the document]
Comment (katspaugh) [reply src/hooks/wallets/useInitWeb3.ts:34]: That doesn't help me. I want to delay the initialization of a read-only provider until a wallet is connected (or it's certain that it won't be autoconnected).

File: src/services/onboard.ts [hunk 3]
```diff
@@ -15,19 +13,15 @@ export type ConnectedWallet = {
...
+export const createOnboard = (chainConfigs: ChainInfo[], currentChain: ChainInfo): OnboardAPI => {
   if (onboard) return onboard
 
   const wallets = getAllWallets(currentChain)
 
   const chains = chainConfigs.map((cfg) => ({
     id: hexValue(parseInt(cfg.chainId)),
     label: cfg.chainName,
-    rpcUrl: rpcConfig?.[cfg.chainId] || getRpcServiceUrl(cfg.rpcUri),
+    rpcUrl: 'data:application/json;{}',
```
[the remark this answers is not in the document]
Comment (katspaugh) [reply src/services/onboard.ts]: I've reverted that.

File: src/hooks/wallets/useOnboard.ts [hunk 4]
```diff
@@ -24,14 +22,10 @@ export type ConnectedWallet = {
 
 const { getStore, setStore, useStore } = new ExternalStore<OnboardAPI>()
 
-export const initOnboard = async (
-  chainConfigs: ChainInfo[],
-  currentChain: ChainInfo,
-  rpcConfig: EnvState['rpc'] | undefined,
```
[the remark this answers is not in the document]
Comment (katspaugh) [reply src/hooks/wallets/useOnboard.ts]: Reverted.

File: src/components/tx-flow/SafeTxProvider.tsx [hunk 5]
```diff
@@ -36,6 +37,9 @@ const SafeTxProvider = ({ children }: { children: ReactNode }): ReactElement =>
   const [nonceNeeded, setNonceNeeded] = useState<boolean>(true)
   const [safeTxGas, setSafeTxGas] = useState<number>()
 
+  // Init Safe Core SDK
+  useInitSafeCoreSDK()
```
[the remark this answers is not in the document]
Comment (katspaugh) [reply src/components/tx-flow/SafeTxProvider.tsx]: Reverted.

**Review by iamacook — APPROVED** [review]

**Comment by katspaugh** [comment]
@francovenica please note that there'll be still a lot of Infura because Onboard uses it for Ledger and Trezor even if your wallet is connected.

**Comment by francovenica** [comment]
I still see calls from the read-only rpc set in the safe settgings after being connected with a wallet in the correct network.
These could be caused by Dapps, so you guys tell me if it is expected to see these:

In the home page:
rpc

Opening apps like the Drain app and tx builder:
image

**Pull request merged**